splitgpg2/__init__.py

```diff
@@ -819,44 +838,51 @@ def escape_D(data):
     # we send the reserialized form this should be safe.
 
     @classmethod
-    def parse_sexpr(klass, untrusted_arg):
+    def parse_sexpr(klass, untrusted_arg: bytes):
+        if type(untrusted_arg) is not bytes:
+            raise TypeError("invalid type in parse_sexpr")
         if len(untrusted_arg) == 0:
             raise ValueError("no sexpr")
+        if untrusted_arg[0] != ord('('):
+            raise ValueError("sexpr must start with (")
         sexpr, rest = klass._parse_sexpr(untrusted_arg, 0)
         if len(rest) != 0:
             raise ValueError("garbage at end of sexpr")
         if len(sexpr) != 1:
-            raise ValueError("sexpr top level shold have exactly one element")
+            raise ValueError("sexpr top level shold have exactly one element, not %d" % len(sexpr))
         return sexpr[0]
 
     @classmethod
-    def _parse_sexpr(klass, untrusted_arg, nesting):
-        if len(untrusted_arg) == 0:
-            if nesting > 0:
-                raise ValueError("missing closing parenthesis")
+    def _parse_sexpr(klass, untrusted_arg, depth, parens=0):
+        if depth > 10:
+            raise ValueError("nesting depth too deep")
+        if not untrusted_arg:
+            if parens != 0:
+                raise ValueError("Missing close parentheses")
             return ([], b'')
         elif untrusted_arg[0] == ord(')'):
-            if nesting == 0:
-                return ([], untrusted_arg)
+            if parens < 1:
+                raise ValueError("Unexpected close paremtheses")
             return ([], untrusted_arg[1:].lstrip(b' '))
 
-        if untrusted_arg[0] in range(0x30, 0x40):
+        if 0x30 <= untrusted_arg[0] < 0x40:
             length_s, _, rest = untrusted_arg.partition(b':')
             length = int(length_s, 10)
             if len(rest) < length:
                 raise ValueError("Invalid length")
             value = rest[0:length]
             rest = rest[length:]
         elif untrusted_arg[0] == ord('('):
-            value, rest = klass._parse_sexpr(untrusted_arg[1:], nesting + 1)
+            value, rest = klass._parse_sexpr(untrusted_arg[1:], depth + 1, parens + 1)
         else:
-            m = re.match(rb'\A([0-9a-zA-Z-_]+) ?(.*)\Z', untrusted_arg)
+            m = klass.sexpr_regex.match(untrusted_arg)
```

Comment:
I prefer to have it in one regex instead of splitting the parsing. I assume the `(.*)\Z` is reasonable fast.

Reply:
IIRC I was trying to avoid quadratic behavior on long input lines, but the line length limit should take care of that.